Refine the django.conf module check to see if the settings really are configured

olasd opened the pull request:
Some modules, e.g. hypothesis, load some Django modules for their own fixtures.
This means that the `django.conf` module sometimes gets loaded, even though it's
never used (and the settings are not initialized).

This causes unrelated test failures when pytest_django is installed, as
pytest_django considers that having a loaded django.conf means the settings are
set up and ready to be modified. The Django settings object provides a flag to
check this condition, which we now use.

Add a regression test that mimics what hypothesis did which makes pytest-django
fail.

Closes #599.

blueyed commented on pytest_django/lazy_django.py:
```diff
@@ -15,14 +15,19 @@ def skip_if_no_django():
+    configured flag in the Django settings object if django.conf has already
+    been imported.
+
+    """
```
In case this needs to be fixed anyway, please remove the newline at the end.

blueyed commented:
(Inner) test fails during collection..

This fixes it:

```diff
diff --git i/tests/test_django_settings_module.py w/tests/test_django_settings_module.py
index a7f716d..98b3f8e 100644
--- i/tests/test_django_settings_module.py
+++ w/tests/test_django_settings_module.py
@@ -368,18 +368,24 @@ def test_no_ds_but_django_conf_imported(testdir, monkeypatch):
     testdir.makepyfile("""
         import os
 
-        # line copied from hypothesis/extras/django.py
-        from django.conf import settings as django_settings
-
-        from pytest_django.lazy_django import django_settings_is_configured
-
         def test_django_settings_is_configured():
+            # Line copied from hypothesis/extras/django.py
+            from django.conf import settings as django_settings
+
+            from pytest_django.lazy_django import django_settings_is_configured
+
             assert django_settings_is_configured() is False
 
         def test_env():
+            # Line copied from hypothesis/extras/django.py
+            from django.conf import settings as django_settings
+
             assert 'DJANGO_SETTINGS_MODULE' not in os.environ
 
         def test_cfg(pytestconfig):
+            # Line copied from hypothesis/extras/django.py
+            from django.conf import settings as django_settings
+
             assert pytestconfig.option.ds is None
     """)
     r = testdir.runpytest_subprocess('-s')
```

It still fails then without your patch, but only with `test_django_settings_is_configured`.

Especially for `test_cfg` I wonder if this then still tests the initial issue, assuming that the hypothesis import happens during test collection?!

olasd commented:
> (Inner) test fails during collection..
>
> [...]
> 
> It still fails then without your patch, but only with `test_django_settings_is_configured`.
> 
> Especially for `test_cfg` I wonder if this then still tests the initial issue, assuming that the hypothesis import happens during test collection?!

I've modified my patch to import django.conf.settings and then `del` it so that pytest doesn't try to poke in it. I've also added an inner test to check that `django.conf` is really imported.

I'm guessing the QA check is broken because of the recent flake8 release? At least it's not in code I touched :-)

olasd commented:
Ah, and, obviously in the end, having that patch on pytest_django makes it not interfere with my other tests!

blueyed commented on tests/test_django_settings_module.py:
```diff
@@ -359,6 +359,41 @@ def test_cfg(pytestconfig):
+
+        # Don't let pytest poke into this object, generating a
+        # django.core.exceptions.ImproperlyConfigured
+        del django_settings
```
Wow.. `inspect.isclass` / `isinstance` triggers the ImproperlyConfigured already..

blueyed commented:
1. Created https://github.com/pytest-dev/pytest/issues/4266 for discussion about the `isclass` issue.

2. > obviously in the end, having that patch on pytest_django makes it not interfere with my other tests!

What do you mean?
That my patch did not fix the issue for you?

olasd commented:
>     1. Created [pytest-dev/pytest#4266](https://github.com/pytest-dev/pytest/issues/4266) for discussion about the `isclass` issue.
> 
>     2. > obviously in the end, having that patch on pytest_django makes it not interfere with my other tests!
> 
> 
> What do you mean?
> That my patch did not fix the issue for you?

Ah, no, I did my patch before I saw yours; I just meant that the original issue was indeed fixed :)

blueyed commented:
I see, rebased.
Will merge once it is green.
Thank you!
